### ISSUE TITLE
Carousel does not work in a modal

### ISSUE THREAD

**Issue by mixlplix** [chanan/BlazorStrap#380]
If you take any of your carousel examples and try to put them in a modal, The modal pops up and it looks like the carousel is loaded and ready to go, but the carousel does not auto-play and when you click left or right it crashes the app.

Sample Code:

    <BSModal @ref="ImageCarouselModal">
        <BSModalHeader OnClick="@(() => ImageCarouselModal.Hide())">Exterior Images</BSModalHeader>
        <BSModalBody>
            <BSCarousel NumberOfItems="2">
                <BSCarouselIndicators NumberOfItems="2" />
                <div class="carousel-inner">
                    <BSCarouselItem src="https://energytwoimages.blob.core.windows.net/images/wood_siding.jpg" alt="Wood siding">
                        <BSCarouselCaption CaptionText="Wood siding" HeaderText="Exterior" />
                    </BSCarouselItem>
                    <BSCarouselItem src="https://energytwoimages.blob.core.windows.net/images/tongue_and_groove_siding.jpg" alt="Tongue and Groove siding">
                        <BSCarouselCaption CaptionText="Tongue and Groove siding" HeaderText="Exterior" />
                    </BSCarouselItem>
                </div>
                <BSCarouselControl CarouselDirection="CarouselDirection.Previous" NumberOfItems="2" />
                <BSCarouselControl CarouselDirection="CarouselDirection.Next" NumberOfItems="2" />
            </BSCarousel>
        </BSModalBody>
        <BSModalFooter>
        </BSModalFooter>
    </BSModal>

**Comment by MasterSkriptor** [comment]
I'll have a look at this.

**Comment by gomjabar6** [comment]
I've done some research into this and the issue with modals is that the carousel uses the "ontransitionend" event handler to recognize that the new item has finished transitioning into position. Once this event happens, the carousel resets itself to be ready to move again. 

Unfortunately, if the carousel is hidden from the DOM when the transition animation ends, the "ontransitionend" event is never fired. This leaves the carousel in a locked-up state, not ready to auto transition (via timer) or manually transition (via control handles). 

One fix for this could be to not just "hide" the modal content (as is the current behavior) but to actually dispose the modal content until it is in the shown state. That will reset the carousel each time the modal is re-shown (starts at first slide again) but it will also ensure the carousel never crashes as it is now. 
Here is a rough example you can do today: 

```csharp
<BSModal @ref="ImageCarouselModal">
    <BSModalHeader OnClick="@(() => ImageCarouselModal.Hide())">Exterior Images</BSModalHeader>
    <BSModalBody>
        @if (ImageCarouselModal.IsOpen == true)
        {
            <BSCarousel NumberOfItems="@items.Count">
                <BSCarouselIndicators NumberOfItems="@items.Count" />
                <div class="carousel-inner">
                    @for (int i = 0; i < items.Count; i++)
                    {
                        <BSCarouselItem src="@items[i].Source" alt="@items[i].Alt" />
                    }
                </div>
                <BSCarouselControl CarouselDirection="CarouselDirection.Previous" ActiveIndexChanged="@(newIndex => { return; })" NumberOfItems="@items.Count" />
                <BSCarouselControl CarouselDirection="CarouselDirection.Next" ActiveIndexChanged="@(newIndex => { return; })" NumberOfItems="@items.Count" />
            </BSCarousel>
        }

    </BSModalBody>
    <BSModalFooter>
    </BSModalFooter>
</BSModal>
```



Another fix could be to update the carousel component to know if it is visible or not and pause the timer until it is visible. This would probably require adding JS back into the component (which was a recent removal). I believe Bootstrap did something similar. The benefit of this is that it doesn't waste resources animating when not-shown. 

Perhaps there is another fix where the carousel doesn't use "ontransitionend" at all but instead uses something more blazor-y that will allow the items to animate without needing to be on-screen. I don't know if that exists or not.

**Comment by jbomhold3** [comment]
changes from @gomjabar6 did fix this issue.

**Comment by gomjabar6** [comment]
@jbomhold3 , I don't think that #437  fixed the issue here, the root cause is still the reliance on the "ontransitionend" event which won't fire if the carousel is hidden. 

We need to find a different hook to use that will still fire even if the item is hidden. I'm sure there is something from Blazor that will work, I'll keep looking.

**Comment by jbomhold3** [comment]
Spoke to quickly there. I had dropped a carousel in a modal on the demo was working then let it run for a min it stopped. Looking into it

**Comment by jbomhold3** [comment]
Ok know the fix building it in now

**Comment by gomjabar6** [comment]
Excited to see what you found!

**Comment by jbomhold3** [comment]
crap forgot to dispose of the event

**Comment by jbomhold3** [comment]
Sounds good will be watching for it

**Comment by AzureGulf** [comment]
Just wondering if this is fixed? I'm using BlazorStrap 1.5.101-Preview-6 in a server-side Blazor app (Net5.0, VS2022 preview 4.1) and cannot seem to get the carousel to appear when embedded in a BlazorStrap (or any other) modal.
Any pointers appreciated.

**Comment by AzureGulf** [comment]
Apologies - missed your ModalCarousel.razor in SampleCore - rtfcf :) - it works
